cpp/src/neighbors/ann_mg/ann_mg.cuh

```diff
@@ -0,0 +1,750 @@
+        auto& ann_if                          = ann_interfaces_[rank];
+        const auto& comms                     = resource::get_comms(dev_res);
+        RAFT_CUDA_TRY(cudaSetDevice(dev_id));
+
+        if (rank == clique.root_rank_) {  // root rank
+          uint64_t batch_offset = clique.root_rank_ * n_rows_of_current_batch * n_neighbors;
+          auto d_neighbors      = raft::make_device_matrix_view<IdxT, IdxT, row_major>(
+            in_neighbors.data_handle() + batch_offset, n_rows_of_current_batch, n_neighbors);
+          auto d_distances = raft::make_device_matrix_view<float, IdxT, row_major>(
+            in_distances.data_handle() + batch_offset, n_rows_of_current_batch, n_neighbors);
+          ann_if.search(dev_res,
+                        search_params,
+                        query_partition,
+                        d_neighbors,
+                        d_distances);  // write search results inplace
+
+          // wait for results of other ranks
+          RAFT_NCCL_TRY(ncclGroupStart());
+          for (int from_rank = 0; from_rank < num_ranks_; from_rank++) {
+            if (from_rank == clique.root_rank_) continue;
+
+            batch_offset = from_rank * n_rows_of_current_batch * n_neighbors;
+            comms.device_recv(in_neighbors.data_handle() + batch_offset,
+                              n_rows_of_current_batch * n_neighbors,
+                              from_rank,
+                              resource::get_cuda_stream(dev_res));
+            comms.device_recv(in_distances.data_handle() + batch_offset,
+                              n_rows_of_current_batch * n_neighbors,
+                              from_rank,
+                              resource::get_cuda_stream(dev_res));
+          }
+          RAFT_NCCL_TRY(ncclGroupEnd());
+          resource::sync_stream(dev_res);
+        } else {  // non-root ranks
+          auto d_neighbors = raft::make_device_matrix<IdxT, IdxT, row_major>(
+            dev_res, n_rows_of_current_batch, n_neighbors);
+          auto d_distances = raft::make_device_matrix<float, IdxT, row_major>(
+            dev_res, n_rows_of_current_batch, n_neighbors);
+          ann_if.search(
+            dev_res, search_params, query_partition, d_neighbors.view(), d_distances.view());
+
+          // send results to root rank
+          RAFT_NCCL_TRY(ncclGroupStart());
+          comms.device_send(d_neighbors.data_handle(),
+                            n_rows_of_current_batch * n_neighbors,
+                            clique.root_rank_,
+                            resource::get_cuda_stream(dev_res));
+          comms.device_send(d_distances.data_handle(),
+                            n_rows_of_current_batch * n_neighbors,
+                            clique.root_rank_,
+                            resource::get_cuda_stream(dev_res));
+          RAFT_NCCL_TRY(ncclGroupEnd());
+          resource::sync_stream(dev_res);
+        }
+      }
+#pragma omp barrier
```

Comment:
The parallel region ended a line above. The barrier construct has no effect outside a parallel region.  Please remove this, and all other stray barriers, it makes the code confusing to read.

For ref [OpenMP 5.2](https://www.openmp.org/spec-html/5.2/openmpse57.html): "An implicit barrier occurs at the end of a parallel region."

Reply:
Thanks for noticing this, was a mistake.